src/Data/Text/Builder/Linear/Hex.hs

```diff
@@ -9,33 +9,48 @@ module Data.Text.Builder.Linear.Hex (
 
 import Data.Bits (Bits (..), FiniteBits (..))
 import Data.Text.Array qualified as A
+import Data.Word (Word8, Word16, Word32, Word64)
 import GHC.Exts (Int (..), (>#))
 import GHC.ST (ST)
 
 import Data.Text.Builder.Linear.Core
 
--- | Append the lower-case hexadecimal represensation of a /non-negative/ number.
+-- | Append the lower-case hexadecimal represensation of a number.
 --
--- __Warning:__ this function has an undefined behaviour for negative numbers.
+-- Negative numbers are interpreted as their corresponding unsigned number, e.g.
+--
+-- >>> :set -XOverloadedStrings -XLinearTypes
+-- >>> import Data.Int (Int8, Int16)
+-- >>> runBuffer (\b -> b |>& (-1 :: Int8)) == "ff"
+-- True
+-- >>> runBuffer (\b -> b |>& (-1 :: Int16)) == "ffff"
+-- True
 (|>&) ∷ (Integral a, FiniteBits a) ⇒ Buffer ⊸ a → Buffer
 
 infixl 6 |>&
 buffer |>& n =
   appendBounded
-    (finiteBitSize n `shiftR` 2)
+    (1 + ((finiteBitSize n - 1) `shiftR` 2))
```

Comment:
Let's move this line into a helper function, with `{-# INLINEABLE #-}` pragma.

Reply:
Done